<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Максим! Давай знакомиться! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы». 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
    
Все ключевые этапы в работе выполнены, и я вижу что с проектом ты справшяешься. Есть моменты, которые нужно доработать, но я уверен, у тебя все получится.
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂
 
<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [24]:
import pandas as pd
 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn.dummy import DummyClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

import warnings
warnings.filterwarnings("ignore")

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

In [25]:
df = pd.read_csv('/datasets/users_behavior.csv')
df

calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
...     ...      ...       ...       ...       ...
3209  122.0   910.98      20.0  35124.90         1
3210   25.0   190.36       0.0   3275.61         0
3211   97.0   634.44      70.0  13974.06         0
3212   64.0   462.32      90.0  31239.78         0
3213   80.0   566.09       6.0  29480.52         1

[3214 rows x 5 columns]

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Можно дополнительно посмотреть информацию о данных с помощью метода .info()
</div>

<div class="alert alert-info">
Замечание учтено, правки внесены!
</div>   

<div class="alert alert-info">
Замечание учтено!
</div>   

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

In [27]:
print(f"Процент пользователей с тарифом ULTRA {(df['is_ultra'].sum()/df['is_ultra'].count()):.1f}")
print(f"Процент пользователей с тарифом SMART {(1 - df['is_ultra'].sum()/df['is_ultra'].count()):.1f}")

Процент пользователей с тарифом ULTRA 0.3
Процент пользователей с тарифом SMART 0.7


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Данные загружены и исследованы.</div>

Загрузив датасет можно сказать, что соотношение классов с тарифом `ULTRA` и `SMART` не равные. Примерно **30 процентов** пользователей пользуется тарифом `ULTRA`, а преобладающая часть пользуется тарифом `SMART`, что составляет **70 процентов**, наблюдается разбаланс классов.

## Разбейте данные на выборки

Для обучения и корректной проверки модели нужно разделить исходные данные на обучающую, валидационную и тестовую выборки в пропорции **3:1:1**.

In [28]:
features = df.drop('is_ultra',axis = 1)# извлеките признаки 
target = df['is_ultra']# извлеките целевой признак

features_train, features_check, target_train, target_check \
        = train_test_split(features, target, test_size=0.4, random_state=12345)
features_test, features_valid, target_test, target_valid \
        = train_test_split(features_check, target_check, test_size=0.5, random_state=12345)

print(f"Размер тестовой выборки: {features_test.shape}\
        \nРазмер валидационной выборки: {features_valid.shape}\
        \nРазмер тренировочной выборки: {features_train.shape}") 

Размер тестовой выборки: (643, 4)        
Размер валидационной выборки: (643, 4)        
Размер тренировочной выборки: (1928, 4)


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Выборки получены корректно</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Старайся не делать такие длинные строки. Стандарт PEP8 регулирует длину строки 79 символов. Придерживаться такого очень сложно, но лучше хотя-бы вмещать в экран, чтобы не приходилось использовать scroll.
    
    
А еще  если сомневаешься, как отформатировать правильно - всегда можно найти какой-нибудь оналйн-форматтер, например этот
    
  https://extendsclass.com/python-formatter.html
    
    
    
</div>

<div class="alert alert-info">
Замечание учтено, правки внесены!
</div>   

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец, что вывел размеры полученных выборок, это помогает быстро проверить корректность проведеного разделения.</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Совет:  обрати внимание на аргумент stratify функции train_test_split. Стратификация особенно важна, когда у нас мало данных и возможен дисбаланс классов в обучении / валидации / тесте.
</div>

В результате получается, что тренировочная выборка составляет **60%** от всех данных, что и являлось целью задачи.

## Исследуйте модели

Требуется выбрать самую подходящую модель машинного обучения для задачи классификации - предсказания тарифов пользователей. Для этого рассмотрим подробнее такие модели как: `решающее дерево, случайный лес, логистическая регерессия и "KNeighbors"`.

In [29]:
best_result = 0

for max_depth in range(1, 15):
    modelDTC = DecisionTreeClassifier(random_state=12345, max_depth = (max_depth))
    modelDTC.fit(features_train, target_train)
    test_predictions = modelDTC.predict(features_valid)
    result= accuracy_score(target_valid,test_predictions)

    if result > best_result:
        best_depth = max_depth
        best_model = modelDTC# сохраните наилучшую модель
        best_result = result#  сохраните наилучшее значение метрики accuracy на валидационных данных

print("Accuracy наилучшей модели на валидационной выборке:", best_result,"\nНаилучшая глубина решающего дерева:",best_depth)
modelDTC = best_model

Accuracy наилучшей модели на валидационной выборке: 0.7993779160186625 
Наилучшая глубина решающего дерева: 7


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Дерево решений исследовано  корректно:
    
 - исследовано 14 значений гиперпараметров
 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле
</div>

Для модели решающего дерева был рассмотрен гиперпараметр `max_depth`, были рассмотрены значения от *1 до 15* и выбрана самая лучшая модель. Метрика `Accuracy` данной модели равна **0.799**, доля правильных ответ больше **75%** - данная модель нам подходит.

In [30]:
best_model = None
best_result = 0
for est in range(1, 15):
    modelRC = RandomForestClassifier(random_state=12345, n_estimators=est) # обучите модель с заданным количеством деревьев
    modelRC.fit(features_train,target_train) # обучите модель на тренировочной выборке
    result = modelRC.score(features_valid,target_valid) # посчитайте качество модели на валидационной выборке
    
    if result > best_result:
        best_model = modelRC# сохраните наилучшую модель
        best_result = result#  сохраните наилучшее значение метрики accuracy на валидационных данных

print("Accuracy наилучшей модели на валидационной выборке:", best_result)
modelRC = best_model 

Accuracy наилучшей модели на валидационной выборке: 0.7869362363919129


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Случайный лес исследован корректно:
    
 - исследовано 14 значений гиперпараметров
 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле
</div>

Для модели случайного леса был рассмотрен гиперпараметр `n_estimators`, были рассмотрены значения от **1 до 15** и выбрана самая лучшая модель. Для данной модели метрика `Accuracy` равна **0.787**, доля правильных ответ больше **75%** - данная модель нам также подходит.

In [31]:
result = 0
best_result = 0
for max_num in range(1, 50):
    for regularization_strength in range(1, 15):
        modelLR = LogisticRegression(random_state = 12345, \
                                     solver = 'lbfgs', max_iter = max_num, \
                                     penalty = 'l2', C = regularization_strength) 
        modelLR.fit(features_train, target_train) 
        result = modelLR.score(features_valid, target_valid) 
    
    if result > best_result:
        best_model = modelLR# сохраните наилучшую модель
        best_result = result#  сохраните наилучшее значение метрики accuracy на валидационных данных
print("Accuracy наилучшей модели на валидационной выборке:",best_result)
modelLR = best_model

Accuracy наилучшей модели на валидационной выборке: 0.6982892690513219


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Старайся давайть переменным осмысленные наименования. Код должен читаться как хороший текст на английском языке, тогда и тебе самому будет проще с ним работать и другим читать и поддерживать. 
</div>

<div class="alert alert-info">
Замечание учтено, правки внесены!
</div>   

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Логистическая регрессия исследована корректно.
</div>


Для модели логистической регрессии были взяты гиперпараметры `max_iter` и `C` при *solver = 'lbfgs'* и *penalty = 'l2'*. Были рассмотрены значения `max_iter` от **1 до 50**, а `C` от **1 до 15**, в результате выбрана самая лучшая модель. Метрика `Accuracy` данной модели равна **0.698**, доля правильных ответ в этом случае меньше **75%** - данная модель нам не может подходить.

In [32]:
result = 0
best_result = 0
for neighbors in range(1, 50):
    for power_parameter in range(1, 2):
        modelKN = KNeighborsClassifier(n_neighbors = neighbors, p = power_parameter)
        modelKN.fit(features_train, target_train) 
        result = modelKN.score(features_valid, target_valid) 
    if result > best_result:
        best_model = modelKN# сохраните наилучшую модель
        best_result = result#  сохраните наилучшее значение метрики accuracy на валидационных данных
print("Accuracy наилучшей модели на валидационной выборке:",best_result)
modelKN = best_model

Accuracy наилучшей модели на валидационной выборке: 0.7573872472783826


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Метод ближайших соседей тоже исследован корректно.
</div>


Для модели `KNeighbors` были взяты гиперпараметры `p`(метрики Минковского) и `n_neighbors`(количество соседей). Были рассмотрено количество соседей от **1 до 50**, а также значения p параметра - **l1 или l2**, в результате выбрана самая лучшая модель. Метрика Accuracy для данной модели равна **0.757**, доля правильных ответ больше **75%** - данная модель нам также подходит.

В заключении для получения несмещенной оценки точности итоговой модели выберем `лучшую модель` по результатам проверки на валидационной выброке. В данном случае наилучшие резельтаты показывает модель `решающего дерева` с параметром `max_depth` равным **7**.


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Ты закончил важный этап: исследовал три модели, нашел оптимальные гиперпараметры для каждой. Нужно зафиксировать результат: выбрать лучшую модель (включая гиперпараметры)  и описать свое решение в выводах.</div>

<div class="alert alert-info">
Замечание учтено, правки внесены!
</div>   

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

## Проверьте модель на тестовой выборке

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
К данному этапу лучше подходить с одной выбранной моделью, чтобы избежать искушения вдруг "поменять" модель по  результатам тестирования. Ведь любое использование тестовой выборки для улучшения модели (в том числе выбор лучшей модели) - это нарушение.

Давай еще раз повторим, зачем мы используем каждую выборку:
  - На обучающей выборке мы подбираем параметры моделей (обучаем модели)
  - На валидационной выборке мы подбираем гиперпараметры, а также все остальные части решения, например предобработку данных, отбор признаков, методы борьбы с дисбалансом и т.д.
  - А тестовая выборка появляется только тогда, когда лучшая модель уже выбрана. И с её помощью мы получаем несмещенную оценку точности модели в эксплуатации.
</div>

<div class="alert alert-info">
Замечание учтено, правки внесены!
</div>   

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

Чтобы проверить точность обученной модели была подготовлена `тестовая выборка`, которую модель будет видеть впервые. 

Для проверки была выбрана модель **решающего дерева**, которая показала наилучшие результаты в `предыдущем параграфе` 

Также `обучающая выборка` была объединена с `валидационной` для большей точности.

In [33]:
features_big_train = pd.concat([features_train, features_valid])
target_big_train = pd.concat([target_train, target_valid])

modelDTC = DecisionTreeClassifier(random_state=12345, max_depth = best_depth)
modelDTC.fit(features_big_train, target_big_train)
 
result = modelDTC.score(features_test,target_test)
print("Accuracy DTC модели на валидационной выборке:",result)

Accuracy DTC модели на валидационной выборке: 0.7916018662519441


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Совет:  Сейчас, когда гиперпараметры подобраны и валидационная выборка нам не нужна, мы можем обучить модель на большой выборке (обучающая + валидационная), а тестирование новой модели произвести уже на тестовой выборке.
    
Объединить выборки можно с помощью pd.concat
</div>

<div class="alert alert-info">
Замечание учтено, правки внесены!
</div>   

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

Модель `решающего дерева` показывает наилучшие результаты с долей правильных ответов - **78%**, что подходит по заданию.

Можно сделать вывод, что самая лучшая модель в данном случае это `решающее дерево`.

## (бонус) Проверьте модели на адекватность

In [34]:
modelDC = DummyClassifier(strategy = "most_frequent", random_state = 12345)
modelDC.fit(features_train, target_train)
print("Accuracy случайной модели на тестовой выборке:",modelDC.score(features_test, target_test))

Accuracy случайной модели на тестовой выборке: 0.7060653188180405


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 Молодец, верно применен Dummy Classifier для оценки стратегии предсказания самого частого ответа.</div>

Для проверки модели на адекватность можно ее сравнить с максимально рандомной моделью, результат которая практически не зависит от данных. При равных отношениях классов модель `DummyClassifier` показывала бы *равновероятностный результат* - **0,5** при выбраном `strategy = "most_frequent"`. В случае с данным датасетом, разабаланс классов состовляет **30/70**, следовательно модель будет показывать с большей вероятностью преобладающий класс.
`Accuracy` модели *DC* на тестовой выборке равно **0.70**, следовательно модели `DTC,RC,KN` работают точнее, чем бросок монеты. Следовательно модели прошли проверку на адекватность.

## Заключение

В данном проекте основной задачей являлося построения системы, способной проанализировать поведение клиентов и предложить пользователям новый тариф. Заказчиком являлся оператор «Мегалайн», который предоставил данные.

Нужно было разбить данные на выборки: убучающую, тестовую и валидационную. После этого требовалось выбрать самую точную модель для предсказания рекомендации тарифов, точность который была бы выше 75%. В конце выбранная модель была проверена на адекватность.

В результате проверки на тестовой выборке, самыми эффективными моделями оказались случайный лес и решающее дерево и "KNeighbors", со значениями Accuracy соотвественно: **0.782**,**0.782**,**0.779**. 

В результате наилучшей моделью было выбрано `рещающее дерево` с параметром `max_depth` равным **7** и долей правильных ответов на тестовой выборке равной **78%**, что подходит требованиям заказчика.

Была проведена проверка моделей на адекватность, путем сравнения их со случайной моделью `DummyClassifier`. Любая из **трех** выбранных выше моделей точнее, чем случайная более чем на **7 процентов**. 

## Чек-лист готовности проекта

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75


<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий</b>

В целом мне все понравилось: твоя работа выполнена на хорошем уровне с минимальными помарками.
Также работа выглядит аккуратной и хорошо оформленной.
Вижу, что тебе дается python, и инструментарий машинного обучения.
 
Единственный момент, который нужно доработать: прошу тебя выбрать лучшую модель по оценкам на валидационной выборке и протестировать её.
    

Жду твой проект на финальное ревью :)
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий v2</b>

Привет! Проект доработат и стал еще лучше. Отлично!

Проект принят. Поздравляю и желаю дальнейших успехов!</div>